src/main/java/com/shapesecurity/bandolier/es2017/transformations/VariableCollisionResolver.java

```diff
@@ -18,36 +21,97 @@ private VariableCollisionResolver() {
+		public final HashTable<Module, HashTable<Variable, String>> renamingMap;
+
+		public ResolvedResult(@Nonnull HashTable<Module, Module> moduleMap, @Nonnull VariableNameGenerator nameGenerator, @Nonnull HashTable<Module, HashTable<Variable, String>> renamingMap) {
+			this.moduleMap = moduleMap;
+			this.nameGenerator = nameGenerator;
+			this.renamingMap = renamingMap;
 		}
-		HashTable<String, HashTable<String, String>> renamingMaps = modules.foldLeft((acc, module) -> acc.put(module.left, HashTable.<String, String>emptyUsingEquality().put("*default*", nameGenerator.next())), HashTable.emptyUsingEquality());
-		for (Pair<String, ImmutableSet<String>> pair : globalNames) {
-			ImmutableSet<String> through = throughNames.get(pair.left).orJust(ImmutableSet.emptyUsingEquality());
-			if (pair.right.length() > 1) {
-				renamingMaps = pair.right.foldAbelian((str, acc) -> through.contains(str) ? acc : acc.put(str, acc.get(str).fromJust().put(pair.left, nameGenerator.next())), renamingMaps);
-			}
+	}
+
+	public static ResolvedResult resolveCollisions(@Nonnull HashTable<String, Module> modules) {
+		HashTable<Module, GlobalScope> globalScopes = modules.foldLeft((acc, pair) -> acc.put(pair.right, ScopeAnalyzer.analyze(pair.right)), HashTable.emptyUsingIdentity());
+		HashTable<Module, ScopeLookup> scopeLookups = modules.foldLeft((acc, pair) -> acc.put(pair.right, new ScopeLookup(globalScopes.get(pair.right).fromJust())), HashTable.emptyUsingIdentity());
+		ImmutableSet<String> allNames = modules.foldLeft((acc, module) -> acc.union(VariableReferenceExtractor.extractAllReferencedVariableNames(globalScopes.get(module.right).fromJust())), ImmutableSet.emptyUsingEquality());
+
+		HashTable<String, ImmutableSet<Pair<Module, ImmutableList<Variable>>>> allDeclaredVariables = modules.foldLeft((acc, module) ->
+			VariableDeclarationExtractor.extractAllDeclaredVariables(globalScopes.get(module.right).fromJust()).entries()
+				.foldLeft((subAcc, pair) ->
+					subAcc.put(pair.left, Pair.of(module.right, pair.right)),
+					acc
+				),
+			MultiHashTable.<String, Pair<Module, ImmutableList<Variable>>>emptyUsingEquality()
+		).toHashTable(ImmutableList::uniqByEquality);
+
+		ImmutableSet<Variable> throughVariables = globalScopes.foldLeft((acc, pair) -> acc.union(pair.right.through.foldLeft((subAcc, subPair) -> subAcc.putAll(subPair.right.map(reference -> {
+			ScopeLookup lookup = scopeLookups.get(pair.left).fromJust();
+			return reference.node instanceof IdentifierExpression ?
+				lookup.findVariableReferencedBy((IdentifierExpression) reference.node) :
+				lookup.findVariableReferencedBy((AssignmentTargetIdentifier) reference.node);
+		})), acc)), ImmutableSet.emptyUsingIdentity());
+
+		VariableNameGenerator nameGenerator = new VariableNameGenerator(allNames.union(throughVariables.map(variable -> variable.name)));
+
+		HashTable<Module, HashTable<Variable, String>> renamingMaps =
+			allDeclaredVariables.entries().filter(pair -> pair.right.length() > 1)
```

Comment:
This is pretty hard to read, and also looks like it contains some duplicate code. Can you, for example, factor out the

```java
(pair, subAcc) ->
  subAcc.put(
    pair.left,
    pair.right.foldLeft(
      (variableAcc, renaming) ->
        variableAcc.put(renaming, nameGenerator.next()),
      subAcc.get(pair.left).orJust(HashTable.emptyUsingIdentity())
    )
  ),
```
bit to a static method with a meaningful name?